akka-distributed-data/src/main/scala/akka/cluster/ddata/VersionVector.scala

```diff
@@ -3,21 +3,37 @@
  */
 package akka.cluster.ddata
 
+import scala.collection.immutable
 import java.util.concurrent.atomic.AtomicLong
-
 import scala.annotation.tailrec
 import scala.collection.immutable.TreeMap
-
 import akka.cluster.Cluster
 import akka.cluster.UniqueAddress
+import akka.cluster.UniqueAddress
 
 /**
  * VersionVector module with helper classes and methods.
  */
 object VersionVector {
 
-  val empty: VersionVector = new VersionVector(TreeMap.empty[UniqueAddress, Long])
+  private val emptyVersions: TreeMap[UniqueAddress, Long] = TreeMap.empty
+  val empty: VersionVector = new VersionVector(Left(emptyVersions))
+
   def apply(): VersionVector = empty
+
+  def apply(versions: TreeMap[UniqueAddress, Long]): VersionVector =
+    if (versions.isEmpty) empty
+    else if (versions.size == 1) apply(versions.head._1, versions.head._2)
+    else new VersionVector(Left(versions))
```

Comment:
Evidence, this is what javap says about `class Foo { val versions: List[(String,String)] = Nil; def doIt(): Unit = { val (k,v) = versions.head}}`. Note `32: new`.

```
public void doIt();
    Code:
       0: aload_0
       1: invokevirtual #17                 // Method versions:()Lscala/collection/immutable/List;
       4: invokevirtual #23                 // Method scala/collection/immutable/List.head:()Ljava/lang/Object;
       7: checkcast     #25                 // class scala/Tuple2
      10: astore_2
      11: aload_2
      12: ifnull        66
      15: aload_2
      16: invokevirtual #28                 // Method scala/Tuple2._1:()Ljava/lang/Object;
      19: checkcast     #30                 // class java/lang/String
      22: astore_3
      23: aload_2
      24: invokevirtual #33                 // Method scala/Tuple2._2:()Ljava/lang/Object;
      27: checkcast     #30                 // class java/lang/String
      30: astore        4
      32: new           #25                 // class scala/Tuple2
      35: dup
      36: aload_3
      37: aload         4
      39: invokespecial #37                 // Method scala/Tuple2."<init>":(Ljava/lang/Object;Ljava/lang/Object;)V
      42: astore        5
      44: aload         5
      46: astore_1
      47: aload_1
      48: invokevirtual #28                 // Method scala/Tuple2._1:()Ljava/lang/Object;
      51: checkcast     #30                 // class java/lang/String
      54: astore        6
      56: aload_1
      57: invokevirtual #33                 // Method scala/Tuple2._2:()Ljava/lang/Object;
      60: checkcast     #30                 // class java/lang/String
      63: astore        7
      65: return
      66: new           #39                 // class scala/MatchError
      69: dup
      70: aload_2
      71: invokespecial #42                 // Method scala/MatchError."<init>":(Ljava/lang/Object;)V
      74: athrow
```